### PR TITLE
HPCC-30776 Add support for generic file formats to ECL

### PR DESCRIPTION
**Pull request by dcamper**
Modify ECL lexer/parser to accept a new syntax for specifying file types for reading (DATASET) and writing (OUTPUT).

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: ecl/hql/hqlgram2.cpp
Comment (ghalliday) [inline ecl/hql/hqlgram2.cpp:8998]: minor: You shouldn't need to copy it to a buffer.  

const char * fileFormatStr = fileFormatAtom->queryStr();

should be sufficient.

You could also use the string from getId() and preserve the original case of the string (and use a case-insensitive comparison)

File: ecl/hql/hqlgram2.cpp
Comment (ghalliday) [inline ecl/hql/hqlgram2.cpp:9021]: Best to still set targetAttr() to something innocuous.  

E.g. targetAttr.setExpr(createValue(no_thor, makeNullType())

so that any further productions always get back an expression.

File: ecl/hql/hqlgram2.cpp
Comment (ghalliday) [inline ecl/hql/hqlgram2.cpp:9020]: Also options.release() to avoid leaking memory.

File: ecl/hql/hqlgram2.cpp
Comment (ghalliday) [inline ecl/hql/hqlgram2.cpp:9014]: There are two options here.  If the type is always first you don't need to wrap it in an attribute.  

```
        OwnedHqlExpr fileTypeExp = createConstant(fileFormatStr);
        HqlExprArray args;
        // Make sure the filetype is first in the args list
        args.append(*LINK(fileTypeExpr));
        options.unwindCommaList(args);
        targetAttr.setExpr(createValue(no_filetype, makeNullType(), args));

``` 
(NOTE: unwindCommaList will append to the array.)

or even use an attribute instead of a string:

```
        HqlExprArray args;
        // Make sure the filetype is first in the args list
        args.append(*createExprAttribute(fileFormatAtom));
        options.unwindCommaList(args);
        targetAttr.setExpr(createValue(no_filetype, makeNullType(), args));
``` 

If you are adding it as an attribute then it doesn't need to go first (although there is no harm if it does).
I would probably opt for 1 of the first 2 items.

**Review by ghalliday — COMMENTED** [review]
@dcamper In general very close to what I would expect.  I have added a few comments - mostly help you know what functions are available, and the most natural way to do something.

File: ecl/hql/hqlgram2.cpp
Comment (dcamper) [reply ecl/hql/hqlgram2.cpp:9014]: I will need to extract the file type name (e.g. "parquet") from the file format options list later on, during code generation.  I am currently doing that by searching for a fileTypeAtom.  If I don't wrap the plugin name in an attribute, how would I know to skip that option for only plugin-provided file types?  (I am hoping this makes sense.)

File: ecl/hql/hqlgram2.cpp
Comment (dcamper) [reply ecl/hql/hqlgram2.cpp:9020]: options is a reference to an attribute generated by the parser.  Where is memory being allocated that needs to be released?

File: ecl/hqlcpp/hqlsource.cpp
Comment (ghalliday) [inline ecl/hqlcpp/hqlsource.cpp:3111]: I think I would approach this as follows:

* Add a parameter forceReadGenerically to SourceBuilder::SourceBuilder and set 
        genericDiskRead = (genericDiskReads && canReadGenerically) || forceReadGenerically;

That means the parameter is set early on, so that anything in the constructor can rely on it.  I also would not worry about checking for hthor - since this is a syntax that the user is explicitly using, and you will get a run time error.

File: ecl/hqlcpp/hqlsource.cpp
Comment (ghalliday) [inline ecl/hqlcpp/hqlsource.cpp:3122]: This is clearer
```
(modeOp == nofiletype) ||
((modeOp == no_csv) && genericDiskRead)
```
Because this code is ensuring something is only done for csv files if they are being read through the generic activity.

File: ecl/hql/hqlgram2.cpp
Comment (ghalliday) [reply ecl/hql/hqlgram2.cpp:9019]: If some options had been supplied.  E.g., TYPE(nonexistant : blah(0))
Then options.expr will contain the expression that corresponds to those options (which is expanded in the options.unwindCommaList() in the true branch.

Destructors are not called on the parser attributes - so if that value isn't used it needs to be freed up explicitly.

**Review by ghalliday — COMMENTED** [review]
Looks close, should be ready to merge soon.

One comment about PRs.  In one case "Best to still set targetAttr() to something innocuous." you made the change and marked the item as resolved.  We have found it is better if the change is in a separate commit - because it makes reviewing easier, and if the reviewer rather than the committer marks the issue as resolved because it provides a reminded for the reviewer to double check the fix.
 
Also, wroth converting from a draft PR, updating the title to  the standard form - which can then be used for the commit message once it is squashed.

**Comment by dcamper** [comment]
@ghalliday My apologies for the 'resolve conversation' snafu.  I do know better but apparently something happened that day.  Either not enough, or too many, migraine drugs, perhaps.

I will address your other points and if that is satisfactory, squash and clean up the PR.

File: ecl/hqlcpp/hqlsource.cpp
Comment (dcamper) [reply ecl/hqlcpp/hqlsource.cpp:3111]: Done.

In parent constructor calls, if DiskReadBuilderBase's canReadGenerically is explicitly false then I pass an explicit false to forceReadGenerically.  Otherwise, I compute the arg value.

File: ecl/hqlcpp/hqlsource.cpp
Comment (dcamper) [reply ecl/hqlcpp/hqlsource.cpp:3122]: Fixed.

File: ecl/hql/hqlgram2.cpp
Comment (dcamper) [reply ecl/hql/hqlgram2.cpp:9019]: Got it, thanks. Added.

**Comment by ghalliday** [comment]
@dcamper for the future, please leave the individual commits and not squash until everything has been signed off.
I will scan the changes again, but it looks close.

**Review by ghalliday — APPROVED** [review]
I think this looks good to merge.  Please can you update the commit message to match the git title, then I will merge.

**Comment by dcamper** [comment]
Commit message fixed.

(How many basic things can I get wrong in one PR?  The world may never know....)

**Comment by ghalliday** [comment]
@dcamper I will delay merging this until I can run a full compiler regression test against it next week - but I don't expect to find anything.

**Pull request merged**
